modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+      this.complete();
+      this.extend(other);
+    }
+
+    pragma "no doc"
+    inline proc deinit() {
+      clear();
+    }
+
+    pragma "no doc"
+    inline proc _destroy(ref item: eltType) {
+      chpl__autoDestroy(item);
+    }
+
+    pragma "no doc"
+    proc _move(ref src: ?t, ref dst: t) {
+      __primitive("=", dst, src);
+    }
+
+    //
+    // The downshift/upshift here (subtract 1 from pos, add 1 to result) is
+    // because this computation assumes zero-based indexing, and Chapel uses
+    // one based indexing.
+    //
+    pragma "no doc"
+    inline proc _getBlockIdx(idx: int): int {
+      const pos = idx + _initialCapacity - 1;
+      const result = log2(pos) - log2(_initialCapacity) + 1;
+      return result;
+    }
+
+    //
+    // Ditto the above.
+    //
+    pragma "no doc"
+    inline proc _getItemIdx(idx: int): int {
+      const pos = idx + _initialCapacity - 1;
+      const result = pos ^ (1 << log2(pos)) + 1;
+      return result;
+    }
+
+    //
+    // Lots of sanity checks for now, just to ensure correctness.
+    //
+    pragma "no doc"
+    proc _getRef(idx: int) ref {
+      _sanity(idx >= 1 && idx <= _totalCapacity);
+      const blockIdx = _getBlockIdx(idx);
+      _sanity(blockIdx >= 1 && blockIdx <= _blocks.capacity);
+      const itemIdx = _getItemIdx(idx);
+      ref block = _blocks[blockIdx];
+      _sanity(block != nil);
+      _sanity(itemIdx >= 1 && itemIdx <= block.capacity);
+      ref result = block[itemIdx];
+      return result;
+    }
+
+    pragma "no doc"
+    inline proc _enter() {
+      if parSafe then
+        _lock$.lock();
+    }
+
+    pragma "no doc"
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+
+    pragma "no doc"
+    inline proc _withinBounds(idx: int): bool {
+      return (idx >= 1 && idx <= _size);
+    }
+
+    //
+    // This call assumes that a lock (acquired by `_enter`) is already
+    // held (if parSafe==true), and releases it before throwing an error if
+    // a bounds check fails.
+    //
+    pragma "no doc"
+    inline proc _boundsCheckLeaveOnThrow(i: int) throws {
+      if !_withinBounds(i) then {
+        _leave();
+        const msg = "List index out of bounds: " + i:string;
+        throw new owned
+          IllegalArgumentError(msg);
+      }
+    }
+
+    pragma "no doc"
+    proc _makeBlockArray(size: int) {
+      type UnmanagedBlock = unmanaged ListBlock;
+      var result = new unmanaged
+        ListBlock(unmanaged ListBlock(eltType), size);
+      return result;
+    }
+
+    //
+    // A call to this assumes that if no free blocks remain, then all free
+    // blocks have been filled and we need to resize the blocks array.
+    //
+    pragma "no doc"
+    proc _maybeResizeBlockArray() {
+
+      const lastBlockIdx = _getBlockIdx(_size);
+      if lastBlockIdx < _blocks.capacity then
+        return;
+
+      var _nblocks = _makeBlockArray(_blocks.capacity * 2);
+      for i in 1.._blocks.capacity do
```

Comment:
Is this really performing an O(N) copy? Why?

Reply:
I see now, this is a `push_back` vector; I thought it was 'substantially similar' to what an unrolled linked list as mentioned here: https://github.com/chapel-lang/chapel/issues/13131#issuecomment-497541624

Reply:
Oh I see now, so `_nblocks[i] = _blocks[i]` is _reusing_ memory from the old blocks, _not_ copying over all elements in the block. So `_blocks` is a block of blocks, and this is resizing the block of blocks so that it can hold more blocks.

Reply:
Right, with the exponential capacity growth, this should be O(log N)